eclipse/che-theia
# Shell layout persistence improvements

### What does this PR do?
This improvement proposal provides refactored mechanism for shell layout persistence. 

Highlighted items extracted from the current proposal change:

- Shell layout save timeout changed from 10 seconds to 2 seconds;
- Removed server side implementation that operated with the storage files and used [FileService](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/filesystem/src/browser/file-service.ts#L271) instead;
- Created an implementation of [FrontendApplicationContribution](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/frontend-application.ts#L33) to allow Che-Theia subscribe to additional window and document events to control UI activity and persist state;
- Created an extension of [ShellLayoutRestorer](https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/shell/shell-layout-restorer.ts#L119) to control the persistence, when Theia goes into `closing_window` state. In this state performing the requests through the websocket can be unexpected, because in this case Theia tries to shut down the registered contributions and websocket channels might be closed;
- Created caching mechanism to prevent storing the same JSON object which was serialized from the application state;
- Removed `che-storage-frontend-contribution.ts` as this implementation implemented in `CheFrontendApplication`;
- Changed event subscription from monitoring add/remove/active widgets to keyboard/mouse events, because this type of events can handle cursor changing position inside opened editor(implementation in master doesn't store cursor position);
- Align with the vanilla Theia store mechanism to persist shell layout across opened workspace roots.

This comes from the provided investigation: https://github.com/eclipse/che/issues/18564#issuecomment-766182346

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### Screenshot/screencast of this PR
State in master branch:
![pannels-folded](https://user-images.githubusercontent.com/1968177/105727952-483fa200-5f34-11eb-98c1-68e2c0068e58.gif)

State within current PR's implementation:
![pannels-restored](https://user-images.githubusercontent.com/1968177/105728027-5ab9db80-5f34-11eb-8658-34155a6c5f7e.gif)

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/18564

### How to test this PR?
Use the following devfile to create the workspace on che.openshift.io:
```yaml
apiVersion: 1.0.0
metadata:
  name: wksp-restore
attributes:
  persistVolumes: 'false'
projects:
  - name: console-java-simple
    source:
      location: 'https://github.com/che-samples/console-java-simple.git'
      type: git
      branch: master
components:
  - type: cheEditor
    reference: 'https://gist.github.com/vzhukovs/732676571880f51bc0b87cc2f7ee1304/raw/c5f007672a49416d66239882da28463f77f96e3b/meta.yaml'
```

After opening the workspace try to wait for 2 seconds and refresh the browser page. State should be restored after page reload.


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable
